### PR TITLE
python 3 port

### PR DESCRIPTION
**Pull request by gforcada**
Most probably it needs some tweaks as python 2 tests run super fast while, for some obscure network interaction on python 3 they take a lot of time.

`mailinglogger.txt` and `summarisinglogger.txt` tests seem to flicker as they may pass or not without a single line of code changed in between...

This pull request is mostly to try it on travis and see what it says.

### CODE REVIEW

File: docs/mailinglogger.txt [hunk 1]
```diff
@@ -140,7 +140,7 @@ sending to ('to@example.com',) from 'from@example.com' using ('localhost', 25)
 Content-Type: text/plain; charset="us-ascii"
 MIME-Version: 1.0
 Content-Transfer-Encoding: 7bit
-Subject: 
+Subject:  
 From: from@example.com
```
Comment (cjw296) [inline docs/mailinglogger.txt]: Why was this needed?

File: docs/summarisinglogger.txt [hunk 2]
```diff
@@ -84,16 +84,26 @@ The logging on script exit is done using python's :mod:`atexit`
...
 
+>>> from mailinglogger.compat import PY2
 >>> import atexit
->>> print atexit._exithandlers
-[(<bound method SummarisingLogger.close of <...>, (), {})]
+>>> if PY2:
+...     len(atexit._exithandlers)
+... else:
+...     atexit._ncallbacks()
+1
```
Comment (cjw296) [inline docs/summarisinglogger.txt]: Well, okay, I see where you're going with this but the docs are there to show that a SummarisingLogger.close has been added as an exit handler, can we get that back?

File: docs/summarisinglogger.txt [hunk 3]
```diff
@@ -84,16 +84,26 @@ The logging on script exit is done using python's :mod:`atexit`
...
-[]
+>>> if PY2:
+...     atexit._exithandlers[:] = []
+... else:
+...     atexit._clear()
+>>> if PY2:
+...     len(atexit._exithandlers)
+... else:
+...     atexit._ncallbacks()
+0
```
Comment (cjw296) [inline docs/summarisinglogger.txt]: See above...

File: mailinglogger/MailingLogger.py [hunk 4]
```diff
@@ -89,8 +89,17 @@ def emit(self, record):
             if self.template is not None:
                 msg = self.template % msg
             subtype = self.content_type.split('/')[-1]
-            if isinstance(msg, unicode):
+            # if possible, ascii encode it.
+            if isinstance(msg, Unicode):
+                try:
+                    msg = msg.encode('ascii')
+                except UnicodeEncodeError:
+                    pass
```
Comment (cjw296) [inline mailinglogger/MailingLogger.py]: Not sure about this block, why was it needed?

File: mailinglogger/MailingLogger.py [hunk 5]
```diff
@@ -89,8 +89,17 @@ def emit(self, record):
...
+
+            if isinstance(msg, Unicode):
                 email = MIMEText(msg, subtype, self.charset)
+            elif isinstance(msg, bytes):
+                email = MIMEText(msg.decode(self.charset), subtype)
```
Comment (cjw296) [inline mailinglogger/MailingLogger.py]: Maybe we should just always ensure unicode and lose the bytes branch here?

File: mailinglogger/SummarisingLogger.py [hunk 6]
```diff
@@ -14,6 +15,8 @@
 
 flood_template = '%i messages not included as flood limit of %i exceeded'
 
+PY3 = sys.version_info[:2] > (3, 0)
+
```
Comment (cjw296) [inline mailinglogger/SummarisingLogger.py]: I'd prefer this to be in .compat and imported wherever it's needed.

File: mailinglogger/SummarisingLogger.py [hunk 7]
```diff
@@ -108,15 +111,20 @@ def close(self):
                 FileHandler.emit(self, record)
 
         FileHandler.close(self)
-        f = os.fdopen(self.fd)
-        summary = f.read().decode(self.charset)
+        if PY3:
+            f = os.fdopen(self.fd, encoding=self.charset)
+        else:
+            f = os.fdopen(self.fd)
+        summary = f.read()
```
Comment (cjw296) [inline mailinglogger/SummarisingLogger.py]: I don't follow this change. `f.read()` returns bytes in both 2 and 3, why can't the original line stay?

File: mailinglogger/SummarisingLogger.py [hunk 8]
```diff
@@ -108,15 +111,20 @@ def close(self):
...
+            except UnicodeEncodeError:
+                # unicode it is then
+                pass
         f.close()
-        # try and encode in ascii, to keep emails simpler:
-        try:
-            summary = summary.encode('ascii')
-        except UnicodeEncodeError:
-            # unicode it is then
-            pass
```
Comment (cjw296) [inline mailinglogger/SummarisingLogger.py]: Not sure why this block has moved?

File: mailinglogger/compat.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    # Bytes = bytes
+    # basestring = str
+
+    # class_type_name = 'class'
+    # ClassType = type
+    # exception_module = 'builtins'
+    # new_class = type
+    # self_name = '__self__'
+    # from io import StringIO
+    # xrange = range
```
Comment (cjw296) [inline mailinglogger/compat.py]: Please lose all the commented bits before we merge.

File: docs/mailinglogger.txt (hunk 1 above)
Comment (gforcada) [inline docs/mailinglogger.txt]: That's what nosetest reports me that it expects. Without it, it shows as an error. No idea why an extra space is needed...

File: docs/summarisinglogger.txt (hunk 2 above)
Comment (gforcada) [inline docs/summarisinglogger.txt]: For python 3 you are totally out of luck: http://stackoverflow.com/questions/16042463/how-can-i-get-the-list-of-registered-atexit-functions-in-python3
